File: gherkin/javascript/test/GherkinTest.ts
import assert from 'assert'
import * as gherkin from '../src'
import { Readable } from 'stream'
import { messages } from 'cucumber-messages'

describe('gherkin', () => {
  it('parses gherkin from the file system', async () => {
    const wrappers = await streamToArray(
      gherkin.fromPaths(['testdata/good/minimal.feature'])
    )
    assert.strictEqual(wrappers.length, 3)
  })

  it('parses gherkin from STDIN', async () => {
    const source = messages.Source.fromObject({
      uri: 'test.feature',
      data: `Feature: Minimal

  Scenario: minimalistic
    Given the minimalism
`,
      media: messages.Media.fromObject({
        encoding: 'UTF8',
        contentType: 'text/x.cucumber.gherkin+plain',
      }),
    })

    const envelopes = await streamToArray(gherkin.fromSources([source]))
    assert.strictEqual(envelopes.length, 3)
  })

  it('parses gherkin using the provided default language', async () => {
    const source = messages.Source.fromObject({
      uri: 'test.feature',
      data: `Fonctionnalité: i18n support
  Scénario: Support des caractères spéciaux
    Soit un exemple de scénario en français
`,
      media: messages.Media.fromObject({
        encoding: 'UTF-8',
        contentType: 'text/x.cucumber.gherkin+plain',
      }),
    })
    const wrappers = await streamToArray(
      gherkin.fromSources([source], { defaultDialect: 'fr' })
    )
    assert.strictEqual(wrappers.length, 3)
  })

  it('outputs dialects', async () => {
    const result = gherkin.dialects()
    assert.strictEqual(result.en.name, 'English')
  })
})

async function streamToArray(
  readableStream: Readable
): Promise<messages.IEnvelope[]> {
  return new Promise<messages.IEnvelope[]>(
    (
      resolve: (wrappers: messages.IEnvelope[]) => void,
      reject: (err: Error) => void
    ) => {
      const items: messages.IEnvelope[] = []
      readableStream.on('data', items.push.bind(items))
      readableStream.on('error', (err: Error) => reject(err))
      readableStream.on('end', () => resolve(items))
    }
  )
}

File: gherkin/dart/test/Gherkin_test.dart
import 'package:gherkin/Gherkin.dart';
import 'package:gherkin/language.dart';
import 'package:gherkin/messages.dart';
import 'package:test/test.dart';

void main()
{
  final idGenerator = IdGenerator.incrementingGenerator;

  test('Use this in readme', () async {
    final paths = const <String>['test/assets/testdata/good/minimal.feature'];
    final includeSource = false;
    final includeAst = true;
    final includePickles = true;
    final envelopeStream = Gherkin.fromPaths(paths, includeSource, includeAst
        , includePickles, idGenerator);

    final pickleStream = envelopeStream.where(
        (envelope) => envelope.pickle.isNotEmpty );

    final envelope = await pickleStream.first;
    expect('minimalistic', envelope.pickle.name );
  });


  test('Provides access to the ast', () async {
    final paths = const <String>['test/assets/testdata/good/minimal.feature'];
    final includeSource = false;
    final includeAst = true;
    final includePickles = false;
    final envelopes = await Gherkin.fromPaths(paths, includeSource, includeAst
        , includePickles, idGenerator).toList();

    // Get the AST
    var gherkinDocument = envelopes.first.gherkinDocument;

    // Get the Feature node of the AST
    var feature = gherkinDocument.feature;
    expect('Minimal', feature.name);

    // Get the first Scenario node of the Feature node
    var scenario = feature.children.first.scenario;
    expect('minimalistic', scenario.name);
  } );

  test('Provides access to pickles which are compiled from the ast', () async {
    final paths = const <String>['test/assets/testdata/good/scenario_outline.feature'];
    final envelopes = await Gherkin.fromPaths( paths
      , false, false, true, idGenerator).toList();

    // Get the first pickle
    final pickle = envelopes.first.pickle;

    // Get the first step of the pickle
    final step = pickle.steps.first;
    expect('the minimalism', step.text);
  } );

  test('Parses supplied source', () async {
    final envelope = Gherkin.makeSourceEnvelope(
        'Feature: Minimal\n'
        '\n'
        '  Scenario: minimalistic\n'
        '    Given the minimalism\n', 'test.feature');

    final singletonList = <Envelope>[envelope];

    final envelopes = await Gherkin.fromSources(singletonList
        , false, true, false, idGenerator).toList();

    final gherkinDocument = envelopes.first.gherkinDocument;
    final feature = gherkinDocument.feature;
    expect('Minimal', feature.name);
  } );
}

